src/gurt/telemetry.c

```diff
@@ -0,0 +1,1857 @@
+
+		if (node->d_tm_type == D_TM_DIRECTORY) {
+			node = d_tm_convert_node_ptr(cshmemRoot, node->child);
+			while (node) {
+				if (d_tm_type & node->d_tm_type) {
+					nodelist = d_tm_add_node(node,
+								 nodelist);
+					if (!nodelist) {
+						rc = -DER_NOMEM;
+						D_GOTO(failure, rc);
+					}
+					if (!*head)
+						*head = nodelist;
+				}
+				node = d_tm_convert_node_ptr(cshmemRoot,
+							     node->sibling);
+			}
+		} else {
+			if (d_tm_type & node->d_tm_type) {
+				nodelist = d_tm_add_node(node, nodelist);
+				if (!nodelist) {
+					rc = -DER_NOMEM;
+					D_GOTO(failure, rc);
+				}
+				if (!*head)
+					*head = nodelist;
+			}
+		}
+	}
+failure:
+	free(str);
+	return rc;
+}
+
+uint64_t
+d_tm_get_num_objects(uint64_t *cshmemRoot, char *path, int d_tm_type)
+{
+	uint64_t count = 0;
+	struct d_tm_node_t *node = NULL;
+	struct d_tm_node_t *parentNode = NULL;
+	char str[D_TM_MAX_NAME_LEN];
+	char *token;
+	char *rest = str;
+
+	sprintf(str, "%s", path);
+
+	parentNode = d_tm_get_root(cshmemRoot);
+	node = parentNode;
+	if (parentNode) {
+		while ((token = strtok_r(rest, "/", &rest))) {
+			node = d_tm_find_child(cshmemRoot, parentNode, token);
+			if (!node)
+				/** no node was found matching the token */
+				return count;
+			parentNode = node;
+		}
+		if (!node)
+			node = parentNode;
+
+		if (node->d_tm_type == D_TM_DIRECTORY) {
+			node = d_tm_convert_node_ptr(cshmemRoot, node->child);
+			while (node) {
+				if (d_tm_type & node->d_tm_type)
+					count++;
+				node = d_tm_convert_node_ptr(cshmemRoot,
+							     node->sibling);
+			}
+		} else {
+			if (d_tm_type & node->d_tm_type)
+				count++;
+		}
+	}
+	return count;
+}
+
+
+/*
+ * Frees the memory allocated for the given nodeList
+ * that was allocated by d_tm_list()
+ */
+void
+d_tm_list_free(struct d_tm_nodeList_t *nodeList)
+{
+	struct d_tm_nodeList_t *head = NULL;
+
+	while (nodeList) {
+		head = nodeList->next;
+		free(nodeList);
+		nodeList = head;
+	}
+}
+
+/*
+ * Adds a node to an existing nodeList, or creates it if the list is empty.
+ */
+struct d_tm_nodeList_t *
+d_tm_add_node(struct d_tm_node_t *src, struct d_tm_nodeList_t *nodelist)
+{
+	struct d_tm_nodeList_t *list = NULL;
+
+	if (!nodelist) {
+		nodelist = (struct d_tm_nodeList_t *)malloc(
+						sizeof(struct d_tm_nodeList_t));
+		if (nodelist) {
+			nodelist->node = src;
+			nodelist->next = NULL;
+			return nodelist;
+		}
+		return NULL;
+	}
+
+	list = nodelist;
+
+	/** advance to the last node in the list */
+	while (list->next)
+		list = list->next;
+
+	list->next = (struct d_tm_nodeList_t *)malloc(
+						sizeof(struct d_tm_nodeList_t));
+	if (list->next) {
+		list = list->next;
+		list->node = src;
+		list->next = NULL;
+		return list;
+	}
+	return NULL;
+}
+
+/*
+ * Server side function that allocates the shared memory segment for this rank.
+ */
+uint8_t *
+d_tm_allocate_shared_memory(int rank, size_t mem_size)
+{
+	int shmid;
+	key_t key;
+
+	/** create a unique key for this rank */
```

Comment:
I'm going to change this from << to '+'.  The '<<' is fine for small numbers, but doesn't scale.  I'll fix this in the next update.